Vostok.Logging.File/Configuration/SafeSettingsCache.cs

```diff
@@ -0,0 +1,60 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Vostok.Logging.File.Configuration
+{
+    internal class SafeSettingsCache
+    {
+        private static readonly object CooldownGuard = new object();
+
+        private readonly SafeSettingsProvider provider;
+        private readonly TimeSpan ttl = TimeSpan.FromSeconds(1);
+        private readonly bool enabled;
+
+        private volatile object updateCooldown;
+        private volatile Task updateCacheTask = Task.CompletedTask;
+        private volatile FileLogSettings currentSettings;
+
+        public SafeSettingsCache(Func<FileLogSettings> provider)
+        {
+            this.provider = new SafeSettingsProvider(provider);
+
+            var settings = Get();
+
+            enabled = settings.EnableFileLogSettingsCache;
+
+            if (enabled)
+                currentSettings = settings;
+        }
+
+        public FileLogSettings Get()
+        {
+            if (!enabled)
+                return provider.Get();
+
+            RefreshCacheAsync();
+
+            return currentSettings;
+        }
+
+        public void RefreshCache() => RefreshCacheAsync().GetAwaiter().GetResult();
+
+        private Task RefreshCacheAsync()
+        {
+            if (updateCooldown == null && Interlocked.CompareExchange(ref updateCooldown, CooldownGuard, null) == null)
+            {
+                updateCacheTask = new Task(() => currentSettings = provider.Get());
+
+                updateCacheTask
+                   .ContinueWith(_ => Task.Delay(ttl))
+                   .Unwrap()
```

Comment:
What's the point of this unwrap call?

Reply:
`ContinueWith` schedules a continuation to the base task. `UpdateCacheTask`, in our case, so we can't just make .`ContinueWith(_ => Task.Delay(ttl)).ContinueWith(_ => updateCooldown = null))`, because `ContinueWith` does not preserve the sequence.

In order to avoid this problem, we can use either `Unwrap` or nest a continuation like this:
`updateCacheTask.ContinueWith(_ => Task.Delay(ttl).ContinueWith(_ => updateCooldown = null))`, which looked dirty for me.

As far as I can see, the functional difference between `Unwrap` and `Nested` versions is that `Unwrap` version creates additional task => it eats more memory, but I don't see where this can become a problem.